src/validation.cpp

```diff
@@ -2220,28 +2226,33 @@ class ConnectTrace {
     }
 
     void BlockConnected(CBlockIndex* pindex, const std::shared_ptr<const CBlock>& pblock) {
-        blocksConnected.emplace_back(pindex, pblock);
-        conflictedTxs.emplace_back();
+        assert(!awaitingClear);
+        assert(pindex && pblock);
+        blocksConnected.back().pindex = pindex;
+        blocksConnected.back().pblock = pblock;
+        blocksConnected.emplace_back();
+    }
+
+    std::vector<PerBlockConnectTrace> GetBlocksConnected() {
+        if (!blocksConnected.back().pindex) {
+            blocksConnected.pop_back();
+            assert(!awaitingClear);
+            awaitingClear = true;
```

Comment:
Sorry, the intention of making conflictedTxs a shared_ptr is so that it can be easily passed to CValidationInterface in a later patchset, not so that we can avoid copy here, forgot to mention that when I split up the patch set.